### PR TITLE
Custom signature info.

### PR DESCRIPTION
**Pull request by snhenson**
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

This generalises how we obtain information about a signature. Currently we always attempt to extract the public key and digest algorthm from the signature OID. This doesn't work for RSA-PSS which encodes the digest and additional parameters in the SignatureAlgorithmIdentifier. This means (among other things) attempts to verify certificate chains with PSS signatures will fail because we can't determine the security of the signing digest (try reverting the custom PSS signature info and you'll see this in the verify tests).

This is also needed for other certificate types I'm working on (*cough*  Ed25519). It will also be needed for PSS certificate signature support for TLS.

This is only currently used for certificates but it could be extended easily to other structures such as CRLs.

### CODE REVIEW

**Comment by snhenson** [comment]
Update: rebased, done make update and added documentation.

**Comment by levitte** [comment]
It needs another rebase, it seems...  but otherwise, this looks good to me.

**Comment by snhenson** [comment]
OK, rebased and "make update" redone.

File: crypto/include/internal/x509_int.h [hunk 1]
```diff
@@ -37,6 +37,19 @@ struct X509_name_st {
...
+
+struct x509_sig_info_st {
+    /* NID of message digest */
+    int mdnid;
+    /* NID of public key algorithm */
+    int pknid;
+    /* Security bits */
+    int secbits;
+    /* Various flags */
+    int32_t flags;
```
Comment (kaduk) [inline crypto/include/internal/x509_int.h]: uint32_t?

File: crypto/x509/x509_set.c [hunk 2]
```diff
@@ -157,3 +160,76 @@ const X509_ALGOR *X509_get0_tbs_sigalg(const X509 *x)
...
+int X509_SIG_INFO_get(const X509_SIG_INFO *siginf, int *mdnid, int *pknid,
+                      int *secbits, uint32_t *flags)
+{
+    if (mdnid)
+        *mdnid = siginf->mdnid;
+    if (pknid)
+        *pknid = siginf->pknid;
+    if (secbits)
+        *secbits = siginf->secbits;
+    if (flags)
```
Comment (kaduk) [inline crypto/x509/x509_set.c]: != NULL (all four)

File: crypto/rsa/rsa_ameth.c [hunk 3]
```diff
@@ -785,6 +785,40 @@ static int rsa_item_sign(EVP_MD_CTX *ctx, const ASN1_ITEM *it, void *asn,
...
+    /*
+     * For TLS need SHA256, SHA384 or SHA512, digest and MGF1 digest must
+     * match and salt length must equal digest size
+     */
+    if ((mdnid == NID_sha256 || mdnid == NID_sha384 || mdnid == NID_sha512)
+            && mdnid == EVP_MD_type(mgf1md) && saltlen == EVP_MD_size(md))
+        flags = X509_SIG_INFO_TLS;
+    else
+        flags = 0;
+    X509_SIG_INFO_set(siginf, mdnid, EVP_PKEY_RSA_PSS, EVP_MD_size(md) * 4,
```
Comment (kaduk) [inline crypto/rsa/rsa_ameth.c]: Does the "times 4" merit a comment?  (I could go either way.)

**Comment by levitte** [comment]
Cool :-)

File: doc/man3/X509_get0_signature.pod [hunk 4]
```diff
@@ -42,13 +51,31 @@ X509_REQ_get0_signature(), X509_REQ_get_signature_nid()
 X509_CRL_get0_signature() and X509_CRL_get_signature_nid() perform the
 same function for certificate requests and CRLs.
 
+X509_get_signature_info() retrieves information about the signature of
+certificate B<x>. The NID of the signing digest is written to B<*mdnid>
+the public key algorithm to B<*pknid> the effective security bits to
+B<*secbits> and flag details to B<*flags>. Any of the parameters can
```
Comment (kaduk) [inline doc/man3/X509_get0_signature.pod]: Probably should put some commas in for the list.

File: doc/man3/X509_get0_signature.pod [hunk 5]
```diff
@@ -42,13 +51,31 @@ X509_REQ_get0_signature(), X509_REQ_get_signature_nid()
...
+
 =head1 NOTES
 
 These functions provide lower level access to signatures in certificates
 where an application wishes to analyse or generate a signature in a form
 where X509_sign() et al is not appropriate (for example a non standard
 or unsupported format).
 
+The security bits returned by X509_get_signature_info() refers to information
+available from the certificate signature (such as the signing digest) in some
```
Comment (kaduk) [inline doc/man3/X509_get0_signature.pod]: Punctuation before "in some"?

File: include/openssl/x509.h [hunk 6]
```diff
@@ -39,6 +39,13 @@
 extern "C" {
 #endif
 
+
+/* Flags for X509_get_signature_info() */
+/* Signature info is valid */
+# define X509_SIG_INFO_VALID     0x1
+/* Signature is suitable for TLS use */
+# define X509_SIG_INFO_TLS       0x2
```
Comment (kaduk) [inline include/openssl/x509.h]: I might think about adding "_OK" to this flag's name; I'm not sure whether I would end up wanting to actually add it, though.

File: crypto/include/internal/x509_int.h (hunk 1 above)
Comment (snhenson) [inline crypto/include/internal/x509_int.h]: Fixed.

File: crypto/x509/x509_set.c (hunk 2 above)
Comment (snhenson) [inline crypto/x509/x509_set.c]: Also fixed.

File: crypto/rsa/rsa_ameth.c (hunk 3 above)
Comment (snhenson) [inline crypto/rsa/rsa_ameth.c]: Comment added to that and the default version which also has * 4.

File: doc/man3/X509_get0_signature.pod (hunk 4 above)
Comment (snhenson) [inline doc/man3/X509_get0_signature.pod]: Fixed.

File: doc/man3/X509_get0_signature.pod (hunk 5 above)
Comment (snhenson) [inline doc/man3/X509_get0_signature.pod]: Fixed.

**Comment by snhenson** [comment]
Addresed @kaduk review comments. Please reconfirm +1.

**Pull request closed without merging**